libgdf/src/validops.cu

```diff
@@ -81,43 +83,71 @@ size_t count_valid_bits_host(std::vector<gdf_valid_type> const & masks, const in
   return count;
 }
 
-constexpr int BITS_PER_MASK32 = GDF_VALID_BITSIZE * RATIO;
 
+/* --------------------------------------------------------------------------*/
+/** 
+ * @brief Kernel to count the number of set bits in a column's validity buffer
+ *
+ * The underlying buffer type may only be a 1B type, but it is casted to a 4B 
+ * type (valid32_t) such that __popc may be used to more efficiently count the 
+ * number of set bits. This requires handling the last 4B element as a special 
+ * case as the buffer may not be a multiple of 4 bytes.
+ * 
+ * @Param[in] masks32 Pointer to buffer (casted as a 4B type) whose bits will be counted
+ * @Param[in] num_masks32 The number of 4B elements in the buffer
+ * @Param[in] num_rows The number of rows in the column, i.e., the number of bits
+ * in the buffer that correspond to rows
+ * @Param[out] global_count The number of set bits in the range of bits [0, num_rows)
+ */
+/* ----------------------------------------------------------------------------*/
+template <typename size_type>
 __global__ 
-void count_valid_bits(valid32_t const * const __restrict__ masks32, 
-                      const int num_masks32, 
-                      const int num_rows, 
-                      int * const __restrict__ global_count)
+void count_valid_bits(valid32_t const * const masks32, 
+                      int const num_masks32, 
+                      int const num_rows, 
+                      size_type * const global_count)
 {
+  using BlockReduce = cub::BlockReduce<size_type, block_size>;
+  __shared__ typename BlockReduce::TempStorage temp_storage;
 
-  int cur_mask = threadIdx.x + blockIdx.x * blockDim.x;
+  // If the number of rows is not a multiple of 32, then the remaining 
+  // rows need to be handled separtely because not all of its bits correspond
+  // to rows
+  int last_mask32{0};
+  const int num_rows_last_mask{num_rows % BITS_PER_MASK32};
+  if(0 == num_rows_last_mask)
+    last_mask32 = num_masks32;
+  else
+    last_mask32 = num_masks32 - 1;
 
-  typedef cub::BlockReduce<int, block_size> BlockReduce;
-  __shared__ typename BlockReduce::TempStorage temp_storage;
+  int const idx{static_cast<int>(threadIdx.x + blockIdx.x * blockDim.x)};
+
+  int cur_mask{idx};
+
+  size_type my_count{0};
 
-  int my_count = 0;
-  while(cur_mask < (num_masks32 - 1))
+  // Use popc to count the valid bits for the all of the masks 
+  // where all of the bits correspond to rows
+  while(cur_mask < last_mask32)
   {
     my_count += __popc(masks32[cur_mask]);
     cur_mask += blockDim.x * gridDim.x;
   }
 
-  if(cur_mask == (num_masks32 - 1))
+  // Handle the remainder rows
+  if(idx < num_rows_last_mask)
   {
-    valid32_t last_mask = masks32[num_masks32 - 1];
-    int num_rows_last_mask = num_rows % BITS_PER_MASK32;
-    if(num_rows_last_mask == 0)
-      num_rows_last_mask = BITS_PER_MASK32;
+    gdf_valid_type const * const valids{reinterpret_cast<gdf_valid_type const *>(masks32)};
```

Comment:
I thought you were from Minnesota, not West const sin.  

Seriously though, you seem to have switched from const west to east const in this new change, but not entirely consistently. (I prefer good old const west...)

Reply:
I'm definitely in the east const camp.  It's a change I've made within the last year, I haven't quite kicked the bad habit of const west yet, so I try and update it when I can, but sometimes I miss a few.

Consider, which is a simpler and more consistent rule to follow:

> const modifies what is on its left. Unless there is nothing on its left, in which case it modifies what’s on its right.
or
> const modifies what is on its left.

I don't know about you, but I prefer the latter. 

Furthermore, consider this example: 

`
using my_type = char *;

// What is the type of t1? It is char * const, NOT const char*
const my_type t1; 

// If instead we do this, it's much clearer that the type of t2 is char * const
my_type const t2;

`

Finally, west const confuses the rule that you read types in C/C++ from right to left.

`const int * const p` 

`p` is a constant pointer to an integer ... that is constant?

OR

`int const * const p`

`p` is a const pointer to a constant integer.